### PR TITLE
LUCENE-10073: Reduce merging overhead of NRT by using a greater mergeFactor on tiny segments.

### PR DESCRIPTION
**Pull request by jpountz**
#11111

### CODE REVIEW

**Review by mikemccand — APPROVED** [review]
Thanks @jpountz -- tricky!

**Comment by mikemccand** [comment]
@jpountz it looks like this one is super-close, and a nice improvement to TMP's default behavior?

File: lucene/core/src/java/org/apache/lucene/index/TieredMergePolicy.java [hunk 1]
```diff
@@ -495,36 +495,36 @@ private MergeSpecification doFindMerges(
...
-
         final List<SegmentCommitInfo> candidate = new ArrayList<>();
         boolean hitTooLarge = false;
-        long bytesThisMerge = 0;
         for (int idx = startIdx;
             idx < sortedEligible.size()
-                && candidate.size() < mergeFactor
-                && bytesThisMerge < maxMergedSegmentBytes;
+                && candidate.size() < maxMergeAtOnce
+                // We allow merging more that mergeFactor segments together if the merged segment
```
[the remark this answers is not in the document]
Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/index/TieredMergePolicy.java]: Thank you, fixed.

File: lucene/core/src/java/org/apache/lucene/index/TieredMergePolicy.java [hunk 2]
```diff
@@ -495,36 +495,36 @@ private MergeSpecification doFindMerges(
...
-            // NOTE: we continue, so that we can try
-            // "packing" smaller segments into this merge
-            // to see if we can get closer to the max
-            // size; this in general is not perfect since
-            // this is really "bin packing" and we'd have
-            // to try different permutations.
-            continue;
           }
           candidate.add(segSizeDocs.segInfo);
-          bytesThisMerge += segBytes;
```
[the remark this answers is not in the document]
Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/index/TieredMergePolicy.java:597]: You are right, these were duplicate variables!

**Comment by jpountz** [comment]
To avoid taking users by surprise, I plan on only changing the default value of maxMergeAtOnce from 10 to 30 in 11.0, meaning that TieredMergePolicy with default parameters will behave exactly the same in 10.x. Users will only experience a different behavior when they update `segsPerTier` or `maxMergeAtOnce` in a way that `maxMergeAtOnce` is greater than `segsPerTier`.

**Comment by jpountz** [comment]
I had missed it, but this change yielded a good speedup on the stored fields benchmark: https://benchmarks.mikemccandless.com/stored_fields_benchmarks.html

**Pull request merged**